solutions/usaco-1159.mdx

```diff
@@ -140,3 +144,145 @@ public class ConnectingTwoBarns {
 </JavaSection>
 
 </LanguageSection>
+
+## Alternative Implementation - DFS + Binary Search
+
+As the editorial mentioned, we can minimize the cost function by using binary search. Since each conncected component array is in sorted order (if we add the nodes of each connected component in increasing order), we can binary search on the sorted connected component array to find the closest field $j$ for each field $i$ using `std::lower_bound`.
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+
+using namespace std;
+using ll = long long;
+using vl = vector<ll>;
+
+#define pb push_back
+
+const ll MAXN = 1e5;
+
+vl adj[MAXN], comps[MAXN];
+ll comp[MAXN];
+
+// DFS to find the connected components
+void dfs(ll cur, ll c)
+{
+	if (comp[cur] != -1)
+	{
+		return;
+	}
+	comp[cur] = c;
+	for (ll u : adj[cur])
+	{
+		dfs(u, c);
+	}
+}
+
+ll cost(ll a, ll b)
+{
+	ll dist = MAXN;
+	for (ll u : comps[a])
+	{
+		/*
+		 * Finds the closest field in bs' connected
+		 * component to field u and updates the minimum
+		 * distance. The closest field is found by binary
+		 * searching on a sorted array (bs' connected 
+		 * component).
+		 */
+		ll i = lower_bound(comps[b].begin(),
+						   comps[b].end(), u) -
+			   comps[b].begin();
+		if (i > 0)
+		{
+			dist = min(dist, abs(comps[b][i - 1] - u));
+		}
+		if (i < comps[b].size())
+		{
+			dist = min(dist, abs(comps[b][i] - u));
+		}
+	}
+	/* 
+	 * Returns the minimum cost of building a path 
+	 * between the two components (aka dist^2)
+	 */
+	return dist * dist;
+}
+
+void solve()
+{
+	ll n, m;
```

Comment:
don't use ll when `int` works just fine

Reply:
too lazy to remove `ll` can u do it urself.